lib/Dialects/LinalgTransform/Transforms/ExpertExpansion.cpp

```diff
@@ -0,0 +1,92 @@
+//===-- ExpertExpansion.cpp - Expand "expert"s to transforms with PDL -----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Dialects/LinalgTransform/LinalgTransformOps.h"
+#include "Dialects/LinalgTransform/Passes.h"
+#include "Dialects/LinalgTransform/SimplePatternRewriter.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassRegistry.h"
+#include "mlir/Rewrite/FrozenRewritePatternSet.h"
+#include "mlir/Rewrite/PatternApplicator.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "expert-expansion"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]")
+
+using namespace mlir;
+using namespace mlir::linalg;
+
+/// Expands the linalg::transform::ExpertOp instances in the `module` into lists
+/// of transformations as described by the `expansions` module that contains
+/// PDL.
+static void expandStrategyOps(ModuleOp module, ModuleOp expansions) {
+  OwningModuleRef clonedExpansions(cast<ModuleOp>(expansions->clone()));
+  RewritePatternSet patterns(std::move(clonedExpansions));
+  FrozenRewritePatternSet frozen(std::move(patterns));
+  PatternApplicator applicator(frozen);
+  applicator.applyDefaultCostModel();
+
+  SimplePatternRewriter rewriter(module.getContext());
+  module.walk([&](transform::ExpertOp expertOp) {
+    rewriter.setInsertionPoint(expertOp);
+    if (failed(applicator.matchAndRewrite(expertOp, rewriter))) {
+      LLVM_DEBUG(DBGS() << "failed to rewrite strategy \""
+                        << expertOp.expertName() << "\"\n");
+    }
+  });
+}
+
+namespace {
+struct ExpertExpansion : public PassWrapper<ExpertExpansion, Pass> {
+  StringRef getArgument() const final {
+    return "linalg-transform-expert-expansion";
+  }
+  StringRef getDescription() const final {
+    return "Expands transformation experts into individual transformations";
+  }
+
+  void runOnOperation() override {
+    auto module = dyn_cast<ModuleOp>(getOperation());
+    if (!module)
+      return signalPassFailure();
+
+    ModuleOp strategyModule = nullptr;
+    for (auto nestedModule : module.getOps<ModuleOp>()) {
+      Optional<StringRef> name = nestedModule.sym_name();
+      if (!name)
+        continue;
+
+      if (*name == "strategies") {
```

Comment:
Can we make this module name an option to the pass?
Not sure how long you want the "strategy" module name to be load bearing but it is magical atm.